File: app/assets/js/save.js
import Rx from 'rx';
import {h} from '@cycle/dom';
import helpers from './helpers'



function view(state$) {
  return state$.map(state => [
      h('.modal.fade#save-modal',
        h('.modal-dialog.modal-md',
          h('.modal-content', [
            h('.modal-header', h('button.close', {attributes: {'data-dismiss': 'modal'}, innerHTML: '&times;'})),
            h('.modal-body', [
              h('p', 'Where do you want to receive the link to your album ?'),
              h('form#save-form', [
                h('.input-group', [
                  h('span.input-group-addon', '@'),
                  h('input.form-control', {type: 'email', name: 'email', placeholder: 'john@doe.com'}),
                  h('span.input-group-btn',
                    h('button.btn.btn-primary#submit-save-btn', 
                      state.status ?
                        [h('i.fa.fa-heart'), ' Saved'] :
                        [h('i.fa.fa-heart-o'), ' Save']
                    )
                  )
                ]),
              ])
            ]),
            h('.modal-footer')
          ]))),
      state.status < 2 ? '' : h('.alert.alert-info.fade.in#save-alert', [h('i.fa.fa-check'), " Saved"])
    ]);
}


module.exports = function(DOM, HTTP, user, collection, album) {

  let actions = {
    modal$: helpers.btn(DOM, '#save-btn'),
    submit$: helpers.btn(DOM, '#submit-save-btn')
      .map(ev => document.getElementById("save-form").elements['email'].value),
    linkSent$: helpers.jsonPOST(HTTP, /\/users\/\d+\/link\/\d+/),
    albumSaved$: helpers.jsonPOSTResponse(HTTP, /\/save/)
  }

  actions.save$ = Rx.Observable.merge(
      album.state$.filter(a => a.length > 1 && !_.some(a, _.isEmpty)),
      collection.state$.skip(2),
      actions.submit$)
    .debounce(2000)

  actions.modal$.subscribe(ev => {
    $('#save-modal').modal('show');
    $('#save-form input[name=email]').focus();
  });

  actions.submit$.subscribe(email => user.actions.updateUser$.onNext({email: email}));


  let state$ = Rx.Observable.merge(
      actions.linkSent$.map(res => state => _.extend(state, {'status': 1})),
      actions.albumSaved$.map(res => state => _.extend(state, {'status': 2})),
      actions.albumSaved$.delay(2000).map(res => state => _.extend(state, {'status': 0}))
    )
    .startWith({'status': 0})
    .scan(helpers.apply)


  let requests = {
    emailLink$: actions.submit$.withLatestFrom(user.state$, collection.state$,
        (email, user, collection) => ({
          url: '/users/' + user.id + '/link/' + collection.hash,
          method: 'POST',
          // eager: true,
          send: {email}
        })),

    saveAlbum$: actions.save$.withLatestFrom(collection.state$, album.state$, (ev, collection, album) => ({
      url: '/save',
      method: 'POST',
      eager: true,
      send: {collection, album}
    })).filter(req => req.send.collection.id && req.send.collection.hash !== helpers.demoID && req.send.album),
  }
  
  return {
    DOM: view(state$),
    HTTP: requests,
    actions,
  }
}

File: client/src/main/scala/bigpiq/components/Save.scala
package bigpiq.client.components

import bigpiq.client._
import org.scalajs.dom
import org.scalajs.dom.raw.HTMLInputElement
import diode.react.ModelProxy
import japgolly.scalajs.react.vdom.prefix_<^._
import japgolly.scalajs.react.{BackendScope, ReactComponentB, _}


object Save {

  val dataToggle = "data-toggle".reactAttr
  val dataTarget = "data-target".reactAttr

  case class Props(proxy: ModelProxy[RootModel])

  case class State()

  class Backend($: BackendScope[Props, State]) {

    val emailInput = Ref[HTMLInputElement]("email")

    def render(props: Props, state: State) = {
      <.div(^.cls := "modal fade", ^.id := "save-modal",
        <.div(^.cls := "modal-dialog modal-sm",
          <.div(^.cls := "modal-content",
            <.div(^.cls := "modal-header"),

            <.div(^.cls := "modal-body",
              <.div(^.cls := "input-group",
                <.span(^.cls := "input-group-addon", "@"),
                <.input(^.cls := "form-control", ^.placeholder := "john@doe.com", ^.ref := "email")
              )
            ),

            <.div(^.cls := "modal-footer",
              <.div(^.cls := "alert alert-info text-left clearfix",
                <.p("We will send you an email with a link to edit your album"),
                <.button(^.cls := "btn btn-primary pull-right",
                  ^.onClick --> props.proxy.dispatch(EmailAndSave(emailInput($).get.value)),
                  "Save"),
                <.button(
                  ^.cls := "btn btn-default pull-right",
                  dataToggle := "modal",
                  dataTarget := "#save-modal",
                  "Cancel")
              )
            )
          )
        )
      )
    }
  }

  def apply(props: Props) = ReactComponentB[Props]("Order")
    .initialState(State())
    .renderBackend[Backend]
    .build
    .apply(props)
}
